src/core/components/pin-set.js

```diff
@@ -0,0 +1,271 @@
+  }
+  const hdrSlice = rootData.slice(vBytes, hdrLength + vBytes)
+  const header = pb.Set.decode(hdrSlice)
+  if (header.version !== 1) {
+    throw new Error(`Unsupported Set version: ${header.version}`)
+  }
+  if (header.fanout > rootNode.links.length) {
+    throw new Error('Impossibly large fanout')
+  }
+  return {
+    header: header,
+    data: rootData.slice(hdrLength + vBytes)
+  }
+}
+
+function hash (seed, key) {
+  const buf = Buffer.alloc(4)
+  buf.writeUInt32LE(seed, 0)
+  const data = Buffer.concat([
+    buf, Buffer.from(toB58String(key))
+  ])
+  return fnv1a(data.toString('binary'))
+}
+
+exports = module.exports = function (dag) {
+  const pinSet = {
+    // should this be part of `object` API?
+    hasChild: (root, childhash, callback) => {
+      let seen = {}
+      if (typeof childhash === 'object') {
+        childhash = toB58String(childhash)
+      }
+
+      function searchChildren (root, cb, _links, _checked) {
+        if (!root.links.length && _links === _checked) {
+          // all nodes have been checked
+          return cb(null, false)
+        }
+
+        some(root.links, (link, cb) => {
+          const bs58Link = toB58String(link.multihash)
+          if (bs58Link in seen) return
+          if (bs58Link === childhash) {
+            return cb(null, true)
+          }
+
+          seen[bs58Link] = true
+
+          dag.get(link.multihash, (err, res) => {
+            if (err) return cb(err)
+
+            _checked++
+            _links += res.value.links.length
+            searchChildren(res.value, cb, _links, _checked)
+          })
+        }, cb)
+      }
+
+      return searchChildren(root, callback, root.links.length, 0)
+    },
+
+    storeSet: (keys, logInternalKey, callback) => {
+      const items = keys.map(key => ({
+        key: key,
+        data: null
+      }))
+
+      pinSet.storeItems(items, logInternalKey, (err, rootNode) => {
+        if (err) { return callback(err) }
+        const opts = { cid: new CID(rootNode.multihash) }
+        dag.put(rootNode, opts, (err, cid) => {
+          if (err) { return callback(err) }
+          logInternalKey(rootNode.multihash)
+          callback(null, rootNode)
+        })
+      })
+    },
+
+    storeItems: (items, logInternalKey, callback, _depth, _subcalls, _done) => {
+      callback = once(callback)
+      const pbHeader = pb.Set.encode({
+        version: 1,
+        fanout: defaultFanout,
+        seed: _depth
+      })
+      let rootData = Buffer.concat([
+        Buffer.from(varint.encode(pbHeader.length)), pbHeader
+      ])
+      let rootLinks = []
+      for (let i = 0; i < defaultFanout; i++) {
+        rootLinks.push(new DAGLink('', 1, emptyKey))
+      }
+      logInternalKey(emptyKey)
+
+      if (items.length <= maxItems) {
+        // the items will fit in a single root node
+        const itemLinks = []
+        const itemData = []
+        const indices = []
+
+        for (let i = 0; i < items.length; i++) {
```

Comment:
I'm not sure about it, but wouldn't it be simpler to have a single array with the items being objects containing `link` and `data` properties? You could then sort them and in the and push them to the corresponding arrays.

Reply:
Yes I initially implemented that but reverted it in favor of keeping the original work. Because you've brought it up, though, I'll refactor it. I have seen some other patterns like this in the codebase so was not sure what others thought of it. I'll look into the sorting question first and then refactor this.